Performance counters (metrics)

bollhals opened the pull request:
## Proposed Changes

This change introduces performance counters for the client.
For now available are:
- total-connection-opened
- open-connection-count
- total-channel-opened
- open-channel-count
- bytes-sent-rate
- bytes-received-rate
- AMQP-method-sent-rate
- AMQP-method-received-rate

fixes the counter part of #776 

I've also fixed in the first commit a bug that happens if the first read into the header buffer returned not the full 7 bytes. (happend a lot when I ran the MassPublish test app with 50 mio messages.)

## Types of Changes

- [x] Bug fix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

bollhals commented:
Do we want other counters or are we fine with the current ones for now?

bollhals commented on projects/RabbitMQ.Client/client/logging/RabbitMqClientEventSource.cs:
```diff
@@ -34,25 +34,20 @@
+        public static readonly RabbitMqClientEventSource Log = new RabbitMqClientEventSource();
+
+        public RabbitMqClientEventSource()
+            : base("rabbitmq-dotnet-client")
```
Do we want it to be named like that? or just "rabbitmq-client"

michaelklishin replied:
This name would not conflict if another client one day provided the same metrics (e.g. Node?)

bollhals replied:
I'd assume it's unlikely you'd use both in the same process, but I actually don't know what would happen if you do have 2 defining the same name. Let me follow up on this

bollhals replied:
(unproven) looks like they would just be merged together probably.

michaelklishin replied:
Then let's go with just "rabbitmq-client".

bollhals replied:
fixed

bollhals commented on projects/RabbitMQ.Client/client/logging/RabbitMqClientEventSource.cs:
```diff
@@ -34,25 +34,20 @@
-    [EventSource(Name="rabbitmq-dotnet-client")]
-    public sealed class RabbitMqClientEventSource : EventSource
+    #nullable enable
+    internal sealed partial class RabbitMqClientEventSource : EventSource
```
I changed it to internal, as others shouldn't be able to do anything with these methods here

michaelklishin commented:
Should the totals be named in plural, e.g.

 * Total connections opened
 * Total channels opened
 * Currently open connection count
 * Currently open channel count

We can also use the terms from the Prometheus/OpenTelemetry communities, e.g. [gauge and counter](https://prometheus.io/docs/concepts/metric_types/) for suffixes.

bollhals commented:
> Should the totals be named in plural, e.g.
> 
> * Total connections opened
> * Total channels opened
> * Currently open connection count
> * Currently open channel count
> 
> We can also use the terms from the Prometheus/OpenTelemetry communities, e.g. [gauge and counter](https://prometheus.io/docs/concepts/metric_types/) for suffixes.

Happy to adapt to whatever we agree upon 👍 
We should probably also consider what already exist in the [dotnet sources](https://source.dot.net/#System.Private.CoreLib/PollingCounter.cs,3c3a2daaa6611183,references)

michaelklishin commented:
Looking at those search results I'd say the names should be

 * Total connections opened
 * Total channels opened
 * Current connection counter
 * Current channel counter

This would very much follow the names in `KestrelEventSource.cs`, for example, but "opened" would be slightly clearer. We can also add a "closed" counterpart so it would be easy to see if there are any connection or channel leaks. "Current" counters do not necessarily make it obvious.

bollhals commented:
I used to have the closed counterpart, but the thing is you always want to know how many are still open, so you have to compare agsinst the open counter. => just have an open/current counter. 
I assume the names eould be the same but the spaces replaced with dashes? (Need both name & displayname)

stebet commented:
This is awesome!

I'd love to see command sent/received as well as frames sent/received if that's not too hard to add as well.

michaelklishin commented:
Yes, names can be underscored or dashed as conventional in this particular community.

bollhals commented:
> This is awesome!
> 
> I'd love to see frames sent/received if that's not too hard to add as well.

Frames received is easy, sent not really as we combine the three frames of a basic deliver method into one byte blob to send. (See the bytes/command sent, they're done in the same place)
OR we'd have to count them before we actually sent them, but this is kind of lying as it wasn't sent yet at that time. 

And only having the frames received is also strange, that's why I havet't added them in this first draft. 

Do you have an idea on how we could solve this best?

bollhals commented:
Updated the names to
- total-connections-opened
- current-open-connections
- total-channels-opened
- current-open-channels

Also updated the event source name to rabbitmq-client

michaelklishin commented:
@bollhals can you please rebase this on top of `master`? Then we can merge.

bollhals commented:
> @bollhals can you please rebase this on top of `master`? Then we can merge.

done, when the build is still successful, we can merge.

xsoheilalizadeh commented:
I'm not able to read these metrics, by dotnet monitor. I'm using Masstranist with Rabbitmq transport, I tried `rabbitmq-client` and `rabbitmq-dotnet-client` for the counter name but It didn't work out. Does anyone have any idea how can I enable this?

bollhals commented:
I'm not sure whether this is in the 6.x branch or is yet to be released.

xsoheilalizadeh commented:
@bollhals When this is going to be merged or released?

lukebakken commented:
@xsoheilalizadeh - I suggest checking out the `master` branch and compiling this library yourself. We didn't plan to back-port this to 6.x.

xsoheilalizadeh commented:
Unfortunately, that's not an option for me.

michaelklishin commented:
@lukebakken given that this is a feature and not a bug, maybe we should ship 7.0 sooner. I wouldn't be against backporting but again, it is not a bug fix and this community cares a lot about us following the strict interpretation of SemVer.

bollhals commented:
Up to you guys. Irrespective pf this, I'd advise to start a thread about 7.0 and the remaining work so anyone has the same understanding and the community knows what/when to expect.

lukebakken commented:
@bollhals please discuss here - https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/1191

xsoheilalizadeh commented:
Do you have any ETA for 7.0 release?

lukebakken commented:
No we do not.

eerhardt commented:
> Also updated the event source name to rabbitmq-client

@bollhals @lukebakken @michaelklishin - is there a reason the event source name was changed from `rabbitmq-dotnet-client` to `rabbitmq-client`? This is a breaking change for anyone who is listening to these EventSource events when moving from 6.x to 7.0.

lukebakken commented:
@eerhardt feel free to open a PR to revert that change. It does seem like an unnecessary break in the API.
